### PR TITLE
UCP/GTEST: make ucx calculate the rndv threshold on its own and test it.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by alinask** [comment]
related to https://github.com/openucx/ucx/issues/493

**Comment by alinask** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1522/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1523/ for details (Mellanox internal link).

**Comment by alinask** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2180/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1524/ for details (Mellanox internal link).

**Comment by alinask** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2181/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1525/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2183/ for details.

**Comment by alinask** [comment]
@MattBBaker Can you please post the error from the ORNL jenkins?

**Comment by alinask** [comment]
@yosefe please review

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -348,11 +348,37 @@ void ucp_ep_config_init(ucp_worker_h worker, ucp_ep_config_t *config)
...
-                config->max_rndv_get_zcopy = iface_attr->cap.get.max_zcopy;
+
+                if (context->config.ext.rndv_thresh == UCS_CONFIG_MEMUNITS_AUTO) {
+                    /* auto */
+                    /* Make UCX calculate the rndv threshold on its own.
+                     * We do it by finding the message size at which rndv and eager_zcopy get
+                     * the same latency. Starting this message size (rndv_thresh), rndv's latency
+                     * would be lower and the reached bandwidth would be higher.
+                     * The used latency functions for eager_zcopy and rndv are specified in the UCX wiki */
+                    rndv_thresh = ((2 * iface_attr->overhead) + (4 * iface_attr->latency)
```
Comment (shamisp) [inline src/ucp/core/ucp_ep.c]: can u document this function in the code ?

**Comment by MattBBaker** [comment]
Looks like a real bug here.
```
[----------] 15 tests from shm/test_ucp_tag_match
[ RUN      ] shm/test_ucp_tag_match.send_recv_nb_partial_exp_medium/0
[       OK ] shm/test_ucp_tag_match.send_recv_nb_partial_exp_medium/0 (102 ms)
[ RUN      ] shm/test_ucp_tag_match.sync_send_unexp/0
[1468926146.318888] [pod9:20047:0]      ib_device.c:132  UCX  WARN  IB Async event on mlx4_0: local access violation work queue error on QPN 0x214a8
[1468926146.318901] [pod9:20047:0]      ib_device.c:132  UCX  WARN  IB Async event on mlx4_0: SRQ-attached QP 0x214a8 was flushed
[1468926146.318919] [pod9:20047:1] rc_verbs_iface.c:51   UCX  ERROR Send completion with error: remote access error
[pod9:20047:1] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x28012)
[pod9:20047:1] Process frozen...
/tmp/hudson3767527985654878374.sh: line 229: 20018 User defined signal 1   $AFFINITY $TIMEOUT make -C test/gtest test UCS_HANDLE_ERRORS=bt GTEST_FILTER="-ud/uct_p2p_am_test.am_zcopy/*:*perf*:test_time.get_time"
Process leaked file descriptors. See http://wiki.jenkins-ci.org/display/JENKINS/Spawning+processes+from+build for more information
Build step 'Execute shell' marked build as failure
make: *** [test] Terminated
```

**Comment by alinask** [comment]
@MattBBaker does this reproduce? 
trying to see if this failure is caused by this PR.
we have several bugs open for similar issues but not sure if they are the same as this one.

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (alinask) [inline src/ucp/core/ucp_ep.c]: yes, will do.

**Comment by MattBBaker** [comment]
I can reproduce this 100% of the time on the test system. I'm attaching a gdb back trace. I'll also test the main branch.
```
0x0000003fcc80f37d in waitpid () from /lib64/libpthread.so.0
#0  0x0000003fcc80f37d in waitpid () from /lib64/libpthread.so.0
#1  0x00007f4768f6a6b4 in ucs_debugger_attach () at debug/debug.c:617
#2  0x00007f4768f6a80d in ucs_error_freeze () at debug/debug.c:662
#3  0x00007f4768f6ae2f in ucs_handle_error () at debug/debug.c:826
#4  0x00007f4768f6addc in ucs_error_signal_handler (signo=11, info=0x7ffd6448fd30, context=0x7ffd6448fc00) at debug/debug.c:808
#5  <signal handler called>
#6  0x00007f4768f66c03 in ucs_mpool_add_to_freelist (mp=0x28012, elem=0x37232f8, add_to_tail=0) at datastruct/mpool.inl:53
#7  0x00007f4768f66c55 in ucs_mpool_put_inline (obj=0x3723300) at datastruct/mpool.inl:67
#8  0x00007f4768f671ab in ucs_mpool_put (obj=0x3723300) at datastruct/mpool.c:150
#9  0x00007f4768d1f06f in uct_rc_txqp_purge_outstanding (txqp=0x37542c8, status=UCS_ERR_ENDPOINT_TIMEOUT, is_log=0) at ib/rc/base/rc_ep.c:430
#10 0x00007f4768d24df6 in uct_rc_verbs_handle_failure (ib_iface=0x35dea20, arg=0x7ffd64490140) at ib/rc/verbs/rc_verbs_iface.c:53
#11 0x00007f4768d2541e in uct_rc_verbs_iface_poll_tx (arg=0x35dea20) at ib/rc/verbs/rc_verbs_iface.c:77
#12 uct_rc_verbs_iface_progress (arg=0x35dea20) at ib/rc/verbs/rc_verbs_iface.c:95
#13 0x00007f4768d15099 in ucs_callbackq_dispatch (cbq=0x3720328) at /ccs/home/bzf/alin/ucx/src/ucs/datastruct/callbackq.h:201
#14 0x00007f4768d1584f in uct_worker_progress (worker=0x3720320) at base/uct_md.c:229
#15 0x00007f47688d055a in ucp_worker_progress (worker=0x3610640) at core/ucp_worker.c:433
#16 0x0000000000537cfc in ucp_test::entity::progress (this=0x3721a80) at ucp/ucp_test.cc:273
#17 0x0000000000535f7b in ucp_test::progress (this=0x371f980) at ucp/ucp_test.cc:68
#18 0x0000000000512b1e in test_ucp_tag::wait (this=0x371f980, req=0x36c3670) at ucp/test_ucp_tag.cc:80
#19 0x00000000005135b0 in test_ucp_tag::recv_b (this=0x371f980, buffer=0x7ffd64490a40, count=8, datatype=8, tag=4919, tag_mask=65535, info=0x7ffd64490a50) at ucp/test_ucp_tag.cc:150
#20 0x00000000004f7520 in test_ucp_tag_match_sync_send_unexp_Test::test_body (this=0x371f980) at ucp/test_ucp_tag_match.cc:349
#21 0x00000000004449fa in ucs::test_base::run (this=0x371f980) at common/test.cc:140
#22 0x0000000000444b87 in ucs::test_base::TestBodyProxy (this=0x371f980) at common/test.cc:166
#23 0x00000000004d1b9a in ucp_test::TestBody (this=0x371f980) at ucp/ucp_test.h:27
#24 0x00000000004341b6 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x371f9d8, method=&virtual testing::Test::TestBody(), location=0x63b193 "the test body") at common/gtest-all.cc:3562
#25 0x000000000042f030 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x371f9d8, method=&virtual testing::Test::TestBody(), location=0x63b193 "the test body") at common/gtest-all.cc:3598
#26 0x00000000004144d8 in testing::Test::Run (this=0x371f9d8) at common/gtest-all.cc:3635
#27 0x0000000000414d6e in testing::TestInfo::Run (this=0x35466e0) at common/gtest-all.cc:3812
#28 0x0000000000415459 in testing::TestCase::Run (this=0x353cb20) at common/gtest-all.cc:3930
#29 0x000000000041c4b3 in testing::internal::UnitTestImpl::RunAllTests (this=0x34df580) at common/gtest-all.cc:5802
#30 0x00000000004355aa in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x34df580, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl *)) 0x41c232 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x63ba70 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3562
#31 0x000000000042ff3e in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x34df580, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl *)) 0x41c232 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x63ba70 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3598
#32 0x000000000041af16 in testing::UnitTest::Run (this=0x950840) at common/gtest-all.cc:5416
#33 0x000000000043de73 in RUN_ALL_TESTS () at common/gtest.h:20058
#34 0x000000000043ddf0 in main (argc=1, argv=0x7ffd64491178) at common/main.cc:79
```

**Comment by yosefe** [comment]
bot:retest

**Comment by yosefe** [comment]
@alinask please check if the ORNL failure is related to rendezvous protocol
@MattBBaker can you please add wiki page which describes ORNL setup (e.g pod9 is OFED  version x, redhat y.x, HCA type, ...)?

**Comment by yosefe** [comment]
@brminich please check why we had segfault in the error handling flow

**Comment by brminich** [comment]
I guess this problem was addressed in UD err handling commit (not merged yet). I'll submit a separate PR for this.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1533/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2192/ for details.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2198/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1538/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2201/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1541/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@MattBBaker is the ORNL jenkins failure the same?

**Comment by MattBBaker** [comment]
Seems similar, it looks like it was taking an error path, though it looks different now.
```
[ RUN      ] shm/test_ucp_tag_match.sync_send_unexp/0
[1469476417.048329] [pod9:29127:0]      ib_device.c:132  UCX  WARN  IB Async event on mlx4_0: local access violation work queue error on QPN 0x17fef
[1469476417.048359] [pod9:29127:0]      ib_device.c:132  UCX  WARN  IB Async event on mlx4_0: SRQ-attached QP 0x17fef was flushed
[1469476417.048365] [pod9:29127:1] rc_verbs_iface.c:51   UCX  ERROR Send completion with error: remote access error
ucp/test_ucp_tag_match.cc:356: Failure
Value of: recv_data
  Actual: 0
Expected: send_data
Which is: 72623859790382856
[pod9:29127:1] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
(Snipped gdb loading details)
#0  0x0000003fcc80f37d in waitpid () from /lib64/libpthread.so.0
#1  0x00007f8ca071e6b4 in ucs_debugger_attach () at debug/debug.c:617
#2  0x00007f8ca071ee1c in ucs_handle_error () at debug/debug.c:819
#3  0x00007f8ca071eddc in ucs_error_signal_handler (signo=11, info=0x7fff92cbaf70, context=0x7fff92cbae40) at debug/debug.c:808
#4  <signal handler called>
#5  0x0000000000000000 in ?? ()
#6  0x00007f8ca0082643 in ucp_request_complete_send (self=0x3cf1c38, arg=0x0) at core/ucp_request.inl:41
#7  ucp_request_release_pending_send (self=0x3cf1c38, arg=0x0) at core/ucp_request.c:81
#8  0x00007f8ca04cb93a in uct_ep_failed_purge (tl_ep=0x3cd85a0, cb=0x7f8ca0082597 <ucp_request_release_pending_send>, arg=0x0) at base/uct_iface.c:235
#9  0x00007f8ca0080d2a in uct_ep_pending_purge (ep=0x3cd9ac0) at /ccs/home/bzf/alin/ucx/src/uct/api/uct.h:1368
#10 ucp_ep_destory_uct_eps (ep=0x3cd9ac0) at core/ucp_ep.c:209
#11 0x00007f8ca0080eae in ucp_ep_destroy (ep=0x3cd9ac0) at core/ucp_ep.c:223
#12 0x00007f8ca00843cc in ucp_worker_destroy_eps (worker=0x3c55860) at core/ucp_worker.c:408
#13 0x00007f8ca0084459 in ucp_worker_destroy (worker=0x3c55860) at core/ucp_worker.c:415
#14 0x000000000053b169 in ucs::handle<ucp_worker*>::release (this=0x3c2c718) at ./common/test_helpers.h:291
#15 0x000000000053a9b9 in ucs::handle<ucp_worker*>::reset (this=0x3c2c718) at ./common/test_helpers.h:249
#16 0x000000000053a522 in ucs::handle<ucp_worker*>::~handle (this=0x3c2c718, __in_chrg=<value optimized out>) at ./common/test_helpers.h:244
#17 0x000000000053ab90 in ucp_test_base::entity::~entity (this=0x3c2c700, __in_chrg=<value optimized out>) at ucp/ucp_test.h:22
#18 0x000000000053abe8 in ucs::ptr_vector_base<ucp_test_base::entity>::release (this=0x3c19930, ptr=0x3c2c700) at ./common/test_helpers.h:201
#19 0x000000000053a0c8 in ucs::ptr_vector_base<ucp_test_base::entity>::clear (this=0x3c19930) at ./common/test_helpers.h:171
#20 0x0000000000537c9e in ucp_test::cleanup (this=0x3c198c0) at ucp/ucp_test.cc:46
#21 0x0000000000513ee4 in test_ucp_tag::cleanup (this=0x3c198c0) at ucp/test_ucp_tag.cc:38
#22 0x00000000004447fa in ucs::test_base::TearDownProxy (this=0x3c198c0) at common/test.cc:122
#23 0x00000000004d1b7a in ucp_test::TearDown (this=0x3c198c0) at ucp/ucp_test.h:63
#24 0x00000000004341b6 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x3c19918, method=&virtual testing::Test::TearDown(), location=0x63d1e1 "TearDown()") at common/gtest-all.cc:3562
#25 0x000000000042f030 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x3c19918, method=&virtual testing::Test::TearDown(), location=0x63d1e1 "TearDown()") at common/gtest-all.cc:3598
#26 0x000000000041451f in testing::Test::Run (this=0x3c19918) at common/gtest-all.cc:3643
#27 0x0000000000414d6e in testing::TestInfo::Run (this=0x3b8b6e0) at common/gtest-all.cc:3812
#28 0x0000000000415459 in testing::TestCase::Run (this=0x3b81b20) at common/gtest-all.cc:3930
#29 0x000000000041c4b3 in testing::internal::UnitTestImpl::RunAllTests (this=0x3b24580) at common/gtest-all.cc:5802
#30 0x00000000004355aa in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x3b24580, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl *)) 0x41c232 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x63dab0 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3562
#31 0x000000000042ff3e in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x3b24580, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl *)) 0x41c232 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x63dab0 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3598
#32 0x000000000041af16 in testing::UnitTest::Run (this=0x953840) at common/gtest-all.cc:5416
#33 0x000000000043de73 in RUN_ALL_TESTS () at common/gtest.h:20058
#34 0x000000000043ddf0 in main (argc=1, argv=0x7fff92cbbb58) at common/main.cc:79
```

**Comment by yosefe** [comment]
@brminich is it with your fix?

**Comment by MattBBaker** [comment]
Also, started a wikipage on pod9 with information about the system. Let me know what other details you need.  https://github.com/openucx/ucx/wiki/POD9

**Comment by brminich** [comment]
@yosefe, yes but the problem looks different now. It fails on user send callback call. 
Why does **shm**/test_ucp_tag_match.sync_send_unexp/0 fail on RC?

**Comment by yosefe** [comment]
@brminich shm includes also IB transport to support access to registered memory (sysv/posix support access only to allocated memory)

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2234/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1567/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@MattBBaker Can you please post the output of ucx_info -d on your host and run the failing test with UCX_LOG_LEVEL=debug and attach the ouput or send it to me?

**Comment by MattBBaker** [comment]
@alinask The output is a bit large, ping me bakermb@ornl.gov and I'll attach logs.

**Comment by alinask** [comment]
@MattBBaker thanks for the output.
according to it, the rndv threshold is calculated to 0 which causes rc to fail.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2311/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1636/ for details (Mellanox internal link).

**Comment by alinask** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2315/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1640/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe please review

**Comment by yosefe** [comment]
:+1: 
@shamisp pls review as well

**Comment by shamisp** [comment]
@yosefe what about @MattBBaker issue ? Why the rndv was calculated as zero ?

**Comment by yosefe** [comment]
@shamisp it happened because of negative factor in the calculation
since now tests passes ORNL jenkins, it probably means the problem is solved. Right @MattBBaker ?

**Comment by shamisp** [comment]
👍

**Pull request merged**